virtcontainers/acrn.go

```diff
@@ -0,0 +1,624 @@
+func (a *acrn) acrnPath() (string, error) {
+	p, err := a.config.HypervisorAssetPath()
+	if err != nil {
+		return "", err
+	}
+
+	if p == "" {
+		p, err = a.arch.acrnPath()
+		if err != nil {
+			return "", err
+		}
+	}
+
+	if _, err = os.Stat(p); os.IsNotExist(err) {
+		return "", fmt.Errorf("ACRN path (%s) does not exist", p)
+	}
+
+	return p, nil
+}
+
+// get the ACRNCTL binary path
+func (a *acrn) acrnctlPath() (string, error) {
+	ctlpath, err := a.config.HypervisorCtlAssetPath()
+	if err != nil {
+		return "", err
+	}
+
+	if ctlpath == "" {
+		ctlpath, err = a.arch.acrnctlPath()
+		if err != nil {
+			return "", err
+		}
+	}
+
+	if _, err = os.Stat(ctlpath); os.IsNotExist(err) {
+		return "", fmt.Errorf("ACRN path (%s) does not exist", ctlpath)
+	}
+
+	return ctlpath, nil
+}
+
+// Logger returns a logrus logger appropriate for logging acrn messages
+func (a *acrn) Logger() *logrus.Entry {
+	return virtLog.WithField("subsystem", "acrn")
+}
+
+func (a *acrn) trace(name string) (opentracing.Span, context.Context) {
+	if a.ctx == nil {
+		a.Logger().WithField("type", "bug").Error("trace called before context set")
+		a.ctx = context.Background()
+	}
+
+	span, ctx := opentracing.StartSpanFromContext(a.ctx, name)
+
+	span.SetTag("subsystem", "hypervisor")
+	span.SetTag("type", "acrn")
+
+	return span, ctx
+}
+
+func (a *acrn) memoryTopology() (Memory, error) {
+	memMb := uint64(a.config.MemorySize)
+
+	return a.arch.memoryTopology(memMb), nil
+}
+
+func (a *acrn) appendImage(devices []Device) ([]Device, error) {
+	imagePath, err := a.config.ImageAssetPath()
+	if err != nil {
+		return nil, err
+	}
+
+	if imagePath != "" {
+		devices, err = a.arch.appendImage(devices, imagePath)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return devices, nil
+}
+
+func (a *acrn) buildDevices(initrdPath string) ([]Device, error) {
+	var devices []Device
+
+	console, err := a.getSandboxConsole(a.id)
+	if err != nil {
+		return nil, err
+	}
+
+	// Add bridges before any other devices. This way we make sure that
+	// bridge gets the first available PCI address.
+	devices = a.arch.appendBridges(devices)
+
+	//Add LPC device to the list of other devices.
+	devices = a.arch.appendLPC(devices)
+
+	devices = a.arch.appendConsole(devices, console)
+
+	if initrdPath == "" {
+		devices, err = a.appendImage(devices)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	devices, err = a.createdummyVirtiBlkDev(devices)
+	if err != nil {
+		return nil, err
+	}
+
+	return devices, nil
+
+}
+
+// setup sets the Acrn structure up.
+func (a *acrn) setup(id string, hypervisorConfig *HypervisorConfig, vcStore *store.VCStore) error {
+	span, _ := a.trace("setup")
+	defer span.Finish()
+
+	err := hypervisorConfig.valid()
+	if err != nil {
+		return err
+	}
+
+	a.id = id
+	a.store = vcStore
+	a.config = *hypervisorConfig
+	a.arch = newAcrnArch(a.config)
+
+	if err = a.store.Load(store.Hypervisor, &a.state); err != nil {
+		//TODO: acrn currently supports only known UUIDs
```

Comment:
raise an issue

Reply:
Raised https://github.com/kata-containers/runtime/issues/1785 to track the issue.